mailbox/api/src/test/java/org/apache/james/mailbox/events/EventBusContract.java

```diff
@@ -71,6 +72,28 @@ default void listenerGroupShouldReceiveEvents() {
         verify(listener, times(1)).event(any());
     }
 
+    @Test
+    default void dispatchShouldNotThrowWhenAGroupListenerFails() {
+        MailboxListener listener = newListener();
+        doThrow(new RuntimeException()).when(listener).event(any());
+
+        eventBus().register(listener, new GroupA());
+
+        assertThatCode(() -> eventBus().dispatch(event, NO_KEYS).block())
+            .doesNotThrowAnyException();
+    }
+
+    @Test
+    default void dispatchShouldNotThrowWhenARegisteredListenerFails() {
```

Comment:
well, I think it's not what we want: we use listeners to make our code modular but if you take the quota example, it could very well be in the main mailbox code, it's just an implementation detail.
If a quota update fails, we probably want the "transaction" to fail (ok, we don't have transaction and we may live with some unsynchronised quota without any problem, but that's just an example) so the caller can "undo" the operation or trigger a process to fix the issue later.

Thinking about that, we can't really undo anything because: we don't know which listener got call before the failure.

The conclusion to that is that we need to accumulate results from listeners, dispatch could have the following signature:

```java
class Result {
  Validation synchronousListenersResult;
  Mono<Validation> asynchronousListenersResult;
}

Result dispatch(Event event, Set<Key> keys)
```

WDYT ?

Reply:
I was thinking of traditional Validation monad but you are right that we also need the list of successful  listeners.
We'd better name it another way because Validation relates to an existing concept. Maybe Report ?

Reply:
However, I do really wonder if it is the caller job to be handling failures.

I would argue that this could be specified by the "registrator".

Idea: When I submit an event to the system, I have no idea of the different par of that system and should not need knowledge of every part of that very system.

I would argue that a `dispatch` caller would most likely *ignore* the validation result...

However, the subscriber might have a better idea of what to do with that verylistener..

IE:

```
Registration register(MailboxListener listener, Group group, ErrorHandler errorHandler);
```

& 

```
class ErrorHandler {
    Set<Class<? extends Exception>> retryOn();
    Duration exponentionalDelayUponRetry();
    int retryCount();
    Duration timeout();
}
```

That way, we will be able to configure these values on a **per listener** basis which IMO makes more sens.

Reply:
Seems a better design yes. I was not comfortable having the dispatcher handle the retry without knowing what to retry and why. Here we can easily design a default ErrorHandler, and specific error handlers in some cases (no idempotence, critical listeners, etc.)

Reply:
We discussed that point with @mbaechler  and we came to the conclusion that we need a deeper understanding to appropriately cover error handling.

The plan is to merge this without error handling & further investigate this as part of next Sprint grooming.

Reply:
See https://github.com/linagora/james-project/pull/2050